leafmap/ai.py

```
@@ -0,0 +1,774 @@
+        print("-" * 60)
+
+    if not execute:
+        return code, None
+
+    if output is None:
+        fd, output = tempfile.mkstemp(suffix=".html", prefix="leafmap_ai_")
+        os.close(fd)
+
+    result = run_safe_code(code, output_path=output)
+    attempt = 0
+    history = []  # 错误历史记忆：累积每次修复的 error/traceback，帮助 LLM 避免重复犯错
```

Ruff (0.16.1)

[warning] Comment contains ambiguous `：` (FULLWIDTH COLON). Did you mean `:` (COLON)?

(RUF003)